### PR TITLE
apply-setters and search-replace README and generated docs

### PR DESCRIPTION
**Pull request by phanimarupaka**
This PR adds the README.md and corresponding generated godocs for apply-setters and search-replace functions.

### CODE REVIEW

**Review by mengqiy — COMMENTED** [review]
Can you please add a section in https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/CONTRIBUTING.md#documentation-changes describing the workflow to re-generating the help text?
I know some parts in that doc appear to be out-of-date. Please ignore them for now. I will update it later.

**Comment by phanimarupaka** [comment]
> Can you please add a section in https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/master/CONTRIBUTING.md#documentation-changes describing the workflow to re-generating the help text?
> I know some parts in that doc appear to be out-of-date. Please ignore them for now. I will update it later.

@mengqiy Sure. Will do it as a separate PR referencing this PR as an example.

File: functions/go/apply-setters/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# apply-setters
```
Comment (mengqiy) [inline functions/go/apply-setters/README.md:1]: I assume this doc has been reviewed in the kpt repo.

File: functions/go/Makefile [hunk 2]
```diff
@@ -117,3 +118,9 @@ func-build: func-verify
 func-push:
 	@echo Pushing image $(CURRENT_FUNCTION)...
 	../../scripts/go-function-release.sh push curated
+
+generate:
+	GO111MODULE=on go get -v github.com/GoogleContainerTools/kpt/mdtogo
+	$(GOBIN)/mdtogo search-replace search-replace/generated --license=none --strategy=cmdDocs
```
Comment (mengqiy) [inline functions/go/Makefile:124]: This is not consistent with the rest of the Makefile.
Ideally we only need to add a function name to this list (https://github.com/GoogleContainerTools/kpt-functions-catalog/blob/1223aa9fa879d8c26c398bee3ef6d394a08002d1/functions/go/Makefile#L20-L26) instead of 2 places.

File: functions/go/apply-setters/README.md (hunk 1 above)
Comment (frankfarzan) [reply functions/go/apply-setters/README.md:1]: I suggest reviewing it again. Also need to think about standardizing the function usage across all 1st party functions. Can you work with @phanimarupaka to define a template readme (Similar to how we have examples/template).

File: functions/go/apply-setters/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    # kpt-set: replicas
+```
+
+2. Input values to setters are provided as key-value pairs -- e.g.
+
+```
+    kpt fn eval apply-setters:unstable -- replicas=3
+```
+
+Control flow
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:38]: Why is this relevant to user of this function?

File: functions/go/apply-setters/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
+# apply-setters
+
+### Overview
+
+<!--mdtogo:Short-->
+
+Apply setter values on resources fields. May set either the complete or partial field value.
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:7]: What's a setter?

Comment (frankfarzan) [inline functions/go/apply-setters/README.md:7]: I have no idea what "complete or partial field value means".

File: functions/go/apply-setters/README.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+<!--mdtogo-->
+
+Setters provide a solution for template-free setting of field values. They are a
+safer alternative to other substitution techniques which do not have the context
+of the structured data. Setters may be invoked to modify the configuration
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:13]: we seem to be using 'resource' and 'configuration' interchangeably. It's pervasive in all our docs and is confusing. As part of auditing command references, we should try to make things consistent.

File: functions/go/apply-setters/README.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+### Synopsis
+
+<!--mdtogo:Long-->
+
+```
+kpt fn eval apply-setters:VERSION [DIR] -- [setter_name=setter_value]
+```
+
+Data model
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:24]: I think you can remove lines 24 to 46, they're not adding much value.

File: functions/go/apply-setters/README.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+Apply setter values on resources fields. May set either the complete or partial field value.
+
+<!--mdtogo-->
+
+Setters provide a solution for template-free setting of field values. They are a
+safer alternative to other substitution techniques which do not have the context
+of the structured data. Setters may be invoked to modify the configuration
+using `apply-setters` function to set values.
+
+### Synopsis
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:16]: Where do you discuss:
- scalar vs list
- interpolation

Keep in mind, this is the comprehensive reference documentation for setters.

File: functions/go/apply-setters/README.md (hunk 1 above)
Comment (mengqiy) [reply functions/go/apply-setters/README.md:1]: Sure

File: functions/go/apply-setters/main.go [hunk 8]
```diff
@@ -25,64 +26,16 @@ func main() {
...
+	cmd.Long = generated.ApplySettersLong
+	cmd.Example = generated.ApplySettersExamples
+
 	if err := cmd.Execute(); err != nil {
 		fmt.Println(err)
 		os.Exit(1)
 	}
 }
 
-func usage() string {
```
Comment (frankfarzan) [inline functions/go/apply-setters/main.go:35]: I'd assume something like

```
usage() {
  return readFromMarkdown(pathToMarkdown);
}
```

In the current formulation, we are not providing this functionality to our SDK users without some pre-processing step. We should keep in mind eventually we want parity between languages as well, so it cannot be specific to Go. Here's usage template used in TS SDK:

https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/master/ts/create-kpt-functions/templates/func.mustache

File: functions/go/apply-setters/README.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+<!--mdtogo:Long-->
+
+```
+kpt fn eval apply-setters:VERSION [DIR] -- [setter_name=setter_value]
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:21]: Related to standardizing: This document doesn't actually describe the function ConfigMap directly.

This also ties functions to kpt CLI (Other funcition usage docs don't do this) whereas a function can technically be invoked by any orchestrator, not just kpt. We should discuss pros/cons of this.

**Review by frankfarzan — CHANGES_REQUESTED** [review]

File: functions/go/apply-setters/README.md (hunk 9 above)
Comment (frankfarzan) [reply functions/go/apply-setters/README.md:21]: This also just shows the imperative way of running it.

I think usage docs should just document the function config and behavior. How to run functions using kpt CLI are out of scope and will be covered in detail in kpt docs.

File: functions/go/Makefile (hunk 2 above)
Comment (phanimarupaka) [reply functions/go/Makefile]: Thanks for the pointer. Added new commit. https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/196/commits/b1398f7e8647d7fbea1a766573db9d62c3700c87

File: functions/go/apply-setters/README.md (hunk 1 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md:1]: It is not.

File: functions/go/apply-setters/README.md (hunk 9 above)
Comment (mengqiy) [inline functions/go/apply-setters/README.md:21]: Image name should be complete: `gcr.io/kpt-fn/apply-setters`

**Comment by mengqiy** [comment]
> Can you work with @phanimarupaka to define a template readme (Similar to how we have examples/template).

Created https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/198

File: functions/go/apply-setters/README.md (hunk 3 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Removed it.

File: functions/go/apply-setters/README.md (hunk 4 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Added description.

File: functions/go/apply-setters/README.md (hunk 5 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Agree. We should use `resource` word to refer to resource files.

File: functions/go/apply-setters/README.md (hunk 6 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Done.

File: functions/go/apply-setters/README.md (hunk 7 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md:16]: Separated scalar vs array values setting. Consumers need not know about interpolation. It is not mentioned even in the error messages. Added example for setting partial field value.

File: functions/go/apply-setters/README.md (hunk 9 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Agree. Updated the doc.

Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Updated.

File: functions/go/apply-setters/main.go (hunk 8 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/main.go:35]: We can implement `make generate` for ts functions as well so that mdtogo generates usage strings in .ts file which can be used by ts files. As the `pathToMarkdown` is not shipped with binary, we need preprocessing to generate language strings from README doc.

File: functions/go/apply-setters/README.md (hunk 4 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Removed.

File: functions/go/Makefile [hunk 10]
```diff
@@ -117,3 +132,7 @@ func-build: func-verify
 func-push:
 	@echo Pushing image $(CURRENT_FUNCTION)...
 	../../scripts/go-function-release.sh push curated
+
+func-generate:
```
Comment (frankfarzan) [inline functions/go/Makefile:136]: this is not `generating the function` just the docs right?

File: functions/go/search-replace/README.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+<!--mdtogo:Long-->
+
+```
+kpt fn eval gcr.io/kpt-fns/search-replace:VERSION [DIR] -- [matcher_name=matcher_value]
+```
+
+#### Matchers
+
+```sh
+by-value
```
Comment (frankfarzan) [inline functions/go/search-replace/README.md:30]: What's this shell block supposed to be? Do you want a list instead?

File: functions/go/apply-setters/main.go (hunk 8 above)
Comment (frankfarzan) [reply functions/go/apply-setters/main.go:35]: Two things here:
1. Ideally, we should provide such functionality to function developers automatically instead of asking them to set this up manually.
2. If possible, it's better to not have a preprocessing step at all (with the complexity of setting up make scripts, etc). Instead, can we just provide library utilities that read the `md` at runtime?

It's fine for us to set up preprocessing steps when we develop the kpt CLI (using Go-specific tooling), but asking function developers (in every supported language) to go through this process requires more thought.

File: functions/go/apply-setters/README.md (hunk 9 above)
Comment (frankfarzan) [reply functions/go/apply-setters/README.md]: These docs still show kpt CLI and talk about imperative vs declarative usage, etc. Did we settle on what's in scope for usage docs?

File: functions/go/apply-setters/README.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+can be declared next to setter names.
+
+```yaml
+apiVersion: v1alpha2
+kind: Kptfile
+metadata:
+  name: my-pkg
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fns/apply-setters:unstable
```
Comment (frankfarzan) [inline functions/go/apply-setters/README.md:33]: You're using `unstable` images throughout the usage doc. Related to whether we want to show kpt CLI usage here.

File: functions/go/Makefile (hunk 10 above)
Comment (phanimarupaka) [reply functions/go/Makefile]: Renamed it to docs-generate

File: functions/go/search-replace/README.md (hunk 11 above)
Comment (phanimarupaka) [reply functions/go/search-replace/README.md:30]: `sh` is not needed. Removed.

File: functions/go/apply-setters/main.go (hunk 8 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/main.go:35]: Parsing can't happen runtime. Binary/image can only include go files and not .md files. So we need a pre-processing step to convert README.md to go/ts files and ship them with binary.

Here are the steps for a function development.
1. Function author clones `kpt-functions-catalog` repo and `cd` into functions/go(or)ts folder.
2. Function author develops a function.
3. Author writes README.md
4. Author runs `make generate` to convert equivalent go/ts files for README.md

This also makes sure that the cli docs are consistent with Github docs. I know there are some additional steps here but let me know if you have a better idea to simplify the process.

File: functions/go/apply-setters/README.md (hunk 9 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: I think usage docs should highlight the recommended workflow(declarative or imperative) the function should be used along with pointing to the alternative approach. In search and replace docs, I just talked about imperative flow as it is highly unlikely it will be used in declarative workflow. Let me know your thoughts.

File: functions/go/apply-setters/README.md (hunk 12 above)
Comment (phanimarupaka) [reply functions/go/apply-setters/README.md:33]: What should I use instead ? Just a place holder ? `gcr.io/kpt-fns/apply-setters:VERSION` ?

File: functions/go/apply-setters/README.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+Discover the names of setters in the Kptfile and declare desired values.
+
+```yaml
+apiVersion: v1alpha2
+kind: Kptfile
+metadata:
+  name: my-pkg
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fns/apply-setters:unstable
```
Comment (mengqiy) [inline functions/go/apply-setters/README.md:80]: Should be kpt-fn

Comment (phanimarupaka) [reply functions/go/apply-setters/README.md]: Thanks.

File: functions/go/apply-setters/main.go (hunk 8 above)
Comment (frankfarzan) [reply functions/go/apply-setters/main.go:35]: Functions can be written outside of the the catalog repo. We have a TS SDK that enables users to create an NPM package from scratch:
https://googlecontainertools.github.io/kpt/guides/producer/functions/ts/develop/

Asking them to fork the catalog repo to inherit some Make targets is a strictly worse UX.
An acceptable solution is to extend the SDK (npm scripts) to generate code from readme, but that functionality needs to be part of the SDK. Otherwise what's the point of having an SDK anyways?

We don't have a proper Go SDK today (just a library), but eventually we need it as well.

I don't feel strongly about pre-generation vs runtime, but the the latter is technically possible. Go binary can read a md file included in the container image. 

@mengqiy Please have tracking issues for the AI(s) here.

File: functions/go/apply-setters/README.md (hunk 12 above)
Comment (frankfarzan) [reply functions/go/apply-setters/README.md]: The usage is tied to a particular version, so treat it like how you treat the source code. The question is what's the process for changing this version as the function is versioned?

@mengqiy This is part of the larger versioning story which also include example docs.

There's another discussion (I made a separate comment but GH comment UI sucks): Dependency on kpt CLI in such usage docs. We need to have a clear guideline on that, and include it in the usage template. Do we want each function to talk about imperative versus declarative workflow? We might want to meet up to finalize these decisions.

File: functions/go/apply-setters/main.go (hunk 8 above)
Comment (mengqiy) [reply functions/go/apply-setters/main.go:35]: https://github.com/GoogleContainerTools/kpt/issues/1642 and https://github.com/GoogleContainerTools/kpt/issues/1641

File: functions/go/apply-setters/README.md (hunk 12 above)
Comment (frankfarzan) [reply functions/go/apply-setters/README.md]: Phani can you add TODOs with the tracking issue. We can't have `unstable` shown to users.

**Review by frankfarzan — APPROVED** [review]
lgtm given we're going to follow up on the open questions
